### PR TITLE
Stable parallel computation of generalized moments

### LINKED ISSUE #78: Stable parallel computation of moments
There probably exist parallel algorithms to compute moments like variance, skewness, etc..

Our current solution for variance is numerically unstable.  It would be nice to replace it with something.  If that something happened to also include general moments that would be really nice.  This task seems  intellectually challenging but somewhat simple from an implementation point of view.  It might be good for an ambitious new contributor.

### PR DESCRIPTION
**Pull request by jcrist**
Adds support for computing generalized moments for a dask array. This results in a more stable computation than that in current master, and allows for computing higher order moments as well.

```python
>>> import dask.array as da
>>> import numpy as np

# Poorly conditioned dataset
>>> a = np.array([1., 2., 3.]*100) + 1e8
>>> d = da.from_array(a, chunks=50)

# numpy answer
>>> a.var()
0.66666666666666663

# current master
>>>  d.var().compute()
0.0

# this PR
>>> d.var().compute()
0.66666666655540463

# Higher order moments
>>> from scipy.stats import moment
>>> da.arange(100, chunks=5).moment(4).compute()
1249583.3625
>>> moment(np.arange(100), 4)
1249583.3625
```

Fixes #78.

### CODE REVIEW

**Comment by jcrist** [comment]
This is a very weird incompatibility bug, I'm not entirely sure what's going on, but this happens.

Python 2:
```python
>>> np.divide(82.5, 10.0, dtype='f8')
8.25
```
Python 3
```python
>>> np.divide(82.5, 10.0, dtype='f8')
8.1999999999999993
```

Not sure if this is a numpy bug or not, numpy version 1.9.1 for both.

**Comment by jcrist** [comment]
Possible solution would be to add a function for python 3 compatibility, as scikit-learn did: https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/fixes.py#L84.

**Comment by mrocklin** [comment]
Whoa, strange error.  compat function seems reasonable (though sad)

File: dask/array/core.py [hunk 1]
```diff
@@ -900,6 +900,37 @@ def var(self, axis=None, dtype=None, keepdims=False, ddof=0):
...
+            array type.
+        keepdims : bool, optional
+            If this is set to True, the axes which are reduced are left in the
+            result as dimensions with size one. With this option, the result
+            will broadcast correctly against the original array.
+        ddof : int, optional
+            "Delta Degrees of Freedom": the divisor used in the calculation is
+            N - ddof, where N represents the number of elements. By default
+            ddof is zero.
+
```
Comment (mrocklin) [inline dask/array/core.py]: Suggest that you include a reference to the tech report you were working from

File: dask/array/reductions.py [hunk 2]
```diff
@@ -12,6 +13,14 @@
 from ..utils import ignoring
 
 
+def getargspec(func):
+    """Version of inspect.getargspec that works for functools.partial objects"""
+    if isinstance(func, partial):
+        return inspect.getargspec(func.func)
+    else:
+        return inspect.getargspec(func)
```
Comment (mrocklin) [inline dask/array/reductions.py]: Toss this into compatibility?

**Comment by mrocklin** [comment]
I'm curious (but not overly concerned) about the performance on naive blocked variance vs this algorithm.

**Comment by jcrist** [comment]
The new chunk function is almost equivalent in workload, but the new aggregate is a bit slower.

Master:
```python
In [1]: import dask.array as da

In [2]: a = da.arange(10000, chunks=20)

In [3]: %timeit a.var().compute()
10 loops, best of 3: 191 ms per loop

In [4]: a = da.arange(10000, chunks=500)

In [5]: %timeit a.var().compute()
100 loops, best of 3: 8.72 ms per loop
```

This PR
```python
In [1]: import dask.array as da

In [2]: a = da.arange(10000, chunks=20)

In [3]: %timeit a.var().compute()
10 loops, best of 3: 199 ms per loop

In [4]: a = da.arange(10000, chunks=500)

In [5]: %timeit a.var().compute()
100 loops, best of 3: 9.2 ms per loop
```

**Comment by mrocklin** [comment]
Oh cool.  I expected a slowdown.  This ended up really nice.

**Comment by mrocklin** [comment]
+1

**Comment by shoyer** [comment]
Just to confirm, this also covers nan-skipping moments like `nanvar`?

**Comment by jcrist** [comment]
nanvar still works, but there's no equivalent nanmoment function. That would be trivial to make though.

**Comment by shoyer** [comment]
Nice, variance covers my use cases.

**Comment by mrocklin** [comment]
@koldunovn might find this PR interesting.  

He is the one that first uncovered that our variance function lost numeric precision.

**Comment by cowlicks** [comment]
Great job!

**Comment by koldunovn** [comment]
Thanks, this is really nice! I will try to test it on my data.

**Pull request merged**